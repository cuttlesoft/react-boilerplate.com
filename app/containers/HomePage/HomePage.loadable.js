import React from 'react'
import loadable from 'utils/loadable'
import { LoadingIndicator } from 'components/LoadingIndicator'

/**
 * Asynchronously loads the component for HomePage
 */
export default loadable(() => import('./HomePage'), {
  fallback: <LoadingIndicator />,
})
