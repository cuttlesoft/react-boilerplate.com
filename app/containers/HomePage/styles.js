import styled from 'styled-components'

export const AtPrefix = styled.span`
  color: black;
  margin-left: 0.4em;
`

export const Form = styled.form`
  margin-bottom: 1em;
`

export const Input = styled.input`
  outline: none;
  border: none;
  border-bottom: 1px dotted #999;
  background-color: transparent;
`

export const Section = styled.section`
  margin: 3em auto;

  &:first-child {
    margin-top: 0;
  }
`
export const CenteredSection = styled(Section)`
  text-align: center;
`
