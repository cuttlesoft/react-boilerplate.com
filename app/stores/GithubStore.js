import { action, observable } from 'mobx'
import { ignore } from 'mobx-sync'
import { createContext } from 'react'
import { getRepos } from '../services/github.service'

export class GitHubStore {
  @ignore @observable repos = []

  @observable currentUser = ''

  @ignore @observable isLoading = false

  @ignore @observable hasError = false

  @action.bound
  async getRepos(username) {
    this.isLoading = true
    this.hasError = false

    try {
      this.repos = await getRepos(username)
      this.currentUser = username
      this.isLoading = false
    } catch (error) {
      this.isLoading = false
      this.hasError = true
    }
  }
}

export const github = new GitHubStore()
export const GitHubStoreContext = createContext(github)
