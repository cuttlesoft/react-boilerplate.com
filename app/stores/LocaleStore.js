import { createContext } from 'react'
import { observable, action } from 'mobx'

import { DEFAULT_LOCALE } from '../i18n'

export class LocaleStore {
  @observable locale = DEFAULT_LOCALE

  @action.bound
  changeLocale(locale) {
    this.locale = locale
  }
}

export const locale = new LocaleStore()
export const LocaleStoreContext = createContext(locale)
