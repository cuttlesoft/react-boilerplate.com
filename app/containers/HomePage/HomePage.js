import React, { useContext, useEffect, useState, memo } from 'react'
import { Helmet } from 'react-helmet'
import { FormattedMessage } from 'react-intl'
import { observer } from 'mobx-react'

// Components
import { Header } from 'components/Header'
import { ReposList } from 'components/ReposList'

// Messages
import messages from './HomePage.messages'

// Styles
import { AtPrefix, CenteredSection, Form, Input, Section } from './HomePage.styles'

// Context / Stores
import { GitHubStoreContext } from '../../stores/GithubStore'

/*
 * HomePage
 *
 * The HomePage renders:
 * - The page title and metadata using Helmet
 * - The page header and description
 * - The search input and repository listing for a given user
 *
 * A user can:
 * - Show Github repositories by username
 *   - When a user inputs a username into the input field then hits the
 *     enter key, the form is submitted and search results are retrieved
 *   - Repositories are displayed using the ReposList component, which handles
 *     loading and error messages
 */
export const HomePage = observer(() => {
  const gitHubStore = useContext(GitHubStoreContext)
  const [searchTerm, setSearchTerm] = useState('')

  /**
   * Prevent the default behavior from occuring (for example, a page refresh),
   * then initiate fetching repositories based on the current `searchTerm`
   *
   * @param {object} evt The sythetic event from firing the function via
   *   `onSubmit`
   */
  const onSubmitForm = evt => {
    evt.preventDefault()
    gitHubStore.getRepos(searchTerm)
  }

  /**
   * When the initial username from the store is not null, set the local state
   * for `searchTerm` to the provided username and initiate fetching
   * repositories
   */
  useEffect(() => {
    if (gitHubStore.currentUser && gitHubStore.currentUser.trim().length > 0) {
      setSearchTerm(gitHubStore.currentUser)
      gitHubStore.getRepos(gitHubStore.currentUser)
    }
  }, [])

  return (
    <article>
      <Helmet>
        <title>Home Page</title>
        <meta name="description" content="A React.js Boilerplate application homepage" />
      </Helmet>

      <CenteredSection>
        <Header level="2" color="brand">
          <FormattedMessage {...messages.startProjectHeader} />
        </Header>
        <p>
          <FormattedMessage {...messages.startProjectMessage} />
        </p>
      </CenteredSection>

      <Section>
        <Header level="2">
          <FormattedMessage {...messages.trymeHeader} />
        </Header>

        <Form data-testid="search-form" onSubmit={onSubmitForm}>
          <label htmlFor="searchTerm">
            <FormattedMessage {...messages.trymeMessage} />

            <AtPrefix>
              <FormattedMessage {...messages.trymeAtPrefix} />
            </AtPrefix>

            <Input
              id="searchTerm"
              type="text"
              placeholder="mxstbr"
              value={searchTerm}
              onChange={e => setSearchTerm(e.target.value)}
            />
          </label>
        </Form>

        <ReposList
          loading={gitHubStore.isLoading}
          error={gitHubStore.hasError}
          repos={gitHubStore.repos}
        />
      </Section>
    </article>
  )
})

export default memo(HomePage)
