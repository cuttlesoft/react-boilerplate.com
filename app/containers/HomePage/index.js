/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 */

import React, { useContext, useEffect, useState, memo } from 'react'
import { Helmet } from 'react-helmet'
import { FormattedMessage } from 'react-intl'
import { observer } from 'mobx-react'

import { Header } from 'components/Header'
import ReposList from 'components/ReposList'
import AtPrefix from './AtPrefix'
import CenteredSection from './CenteredSection'
import Form from './Form'
import Input from './Input'
import Section from './Section'
import messages from './messages'

import { GitHubStoreContext } from '../../stores/GithubStore'

export const HomePage = observer(() => {
  const gitHubStore = useContext(GitHubStoreContext)
  const [searchTerm, setSearchTerm] = useState('')

  const onSubmitForm = evt => {
    if (evt !== undefined && evt.preventDefault) evt.preventDefault()
    gitHubStore.getRepos(searchTerm)
  }

  useEffect(() => {
    // When initial state username is not null, submit the form to load repos
    if (gitHubStore.currentUser && gitHubStore.currentUser.trim().length > 0)
      onSubmitForm(gitHubStore.currentUser)
  }, [])

  return (
    <article>
      <Helmet>
        <title>Home Page</title>
        <meta name="description" content="A React.js Boilerplate application homepage" />
      </Helmet>
      <div>
        <CenteredSection>
          <Header level="2">
            <FormattedMessage {...messages.startProjectHeader} />
          </Header>
          <Header level="2">
            <FormattedMessage {...messages.startProjectMessage} />
          </Header>
        </CenteredSection>
        <Section>
          <Header level="2">
            <FormattedMessage {...messages.trymeHeader} />
          </Header>
          <Form onSubmit={onSubmitForm}>
            <label htmlFor="searchTerm">
              <FormattedMessage {...messages.trymeMessage} />
              <AtPrefix>
                <FormattedMessage {...messages.trymeAtPrefix} />
              </AtPrefix>
              <Input
                id="searchTerm"
                type="text"
                placeholder="mxstbr"
                value={searchTerm}
                onChange={e => setSearchTerm(e.target.value)}
              />
            </label>
          </Form>
          <ReposList
            loading={gitHubStore.isLoading}
            error={gitHubStore.hasError}
            repos={gitHubStore.repos}
          />
        </Section>
      </div>
    </article>
  )
})

export default memo(HomePage)
