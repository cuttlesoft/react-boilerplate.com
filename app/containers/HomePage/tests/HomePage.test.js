/**
 * HomePage Test Criteria
 *
 * A user can:
 * - Show Github repositories by username
 *   - When a user inputs a username into the input field then hits the
 *     enter key, the form is submitted and search results are retrieved
 *     - While the call is loading, a loading indicator should be displayed
 *     - If an error occurs, an error message should be displayed
 *     - If no repositories are found, a not found message should be displayed
 */

// Libraries
import React from 'react'
import 'jest-dom/extend-expect'
import { fireEvent, render, waitForElement } from '@testing-library/react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'mobx-react'
import { RouterStore } from 'mobx-react-router'
import MockAdapter from 'axios-mock-adapter'

// Components / Pages
import { HomePage } from '../HomePage'

// Services
import axios from '../../../services/instance'
import { getReposSuccess } from '../../../services/mocks/github.service.mock'

// Context / Stores
import trunk from '../../../configureStore'
import { GitHubStore, GitHubStoreContext } from '../../../stores/GithubStore'

describe('<HomePage />', () => {
  // Allow stores to be accessed in tests
  let stores = null
  let gitHubStore = null

  const TEST_USER = 'cuttlesoft'

  beforeAll(() => {
    // Create MobX Routing store
    const routingStore = new RouterStore()
    gitHubStore = new GitHubStore()

    stores = {
      routing: routingStore,
    }

    // Initialize the stores
    trunk.init()
  })

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider {...stores}>
        <IntlProvider locale="en">
          <HomePage loading={false} error={false} repos={[]} />
        </IntlProvider>
      </Provider>,
    )
    expect(firstChild).toMatchSnapshot()
  })

  it('should successfully load the repository list when a user inputs a username and submits the form', async () => {
    // Configure Axios to mock a successful call for retrieving repos
    const url = `/users/${TEST_USER}/repos?type=all&sort=updated`
    const axiosMock = new MockAdapter(axios)
    const mockData = getReposSuccess
    axiosMock.onGet(url).reply(200, mockData)

    const getReposSpy = jest.spyOn(gitHubStore, 'getRepos')

    const { getByTestId } = render(
      <Provider {...stores}>
        <IntlProvider locale="en">
          <GitHubStoreContext.Provider value={gitHubStore}>
            <HomePage />
          </GitHubStoreContext.Provider>
        </IntlProvider>
      </Provider>,
    )

    // Ensure that the search term starts out as blank
    const input = document.getElementById('searchTerm')
    expect(input.value).toBe('')

    // Add text to the search field
    fireEvent.change(input, { target: { value: TEST_USER } })
    expect(input.value).toBe(TEST_USER)

    // Submit the form
    fireEvent.submit(getByTestId('search-form'))
    expect(getReposSpy).toHaveBeenCalled()

    // Ensure that the loading indicator is shown
    const loadingNode = await waitForElement(() => getByTestId('repo-list-loading'))
    expect(loadingNode)

    // Ensure that the repo list is shown
    const repoList = await waitForElement(() => getByTestId('repo-list'))
    expect(repoList).toHaveTextContent(getReposSuccess[0].name)
    expect(repoList).toHaveTextContent(getReposSuccess[0].open_issues_count)
  })

  it('should automatically load the repository list when a username is preloaded', async () => {
    // Configure Axios to mock a successful call for retrieving repos
    const url = `/users/${TEST_USER}/repos?type=all&sort=updated`
    const axiosMock = new MockAdapter(axios)
    const mockData = getReposSuccess
    axiosMock.onGet(url).reply(200, mockData)

    // Set the intial state on the store to preload a username
    gitHubStore.currentUser = TEST_USER

    const getReposSpy = jest.spyOn(gitHubStore, 'getRepos')

    const { getByTestId } = render(
      <Provider {...stores}>
        <IntlProvider locale="en">
          <GitHubStoreContext.Provider value={gitHubStore}>
            <HomePage />
          </GitHubStoreContext.Provider>
        </IntlProvider>
      </Provider>,
    )

    const input = document.getElementById('searchTerm')
    expect(input.value).toBe(TEST_USER)
    expect(getReposSpy).toHaveBeenCalled()

    // Ensure that the loading indicator is shown
    const loadingNode = await waitForElement(() => getByTestId('repo-list-loading'))
    expect(loadingNode)

    // Ensure that the repo list is shown
    const repoList = await waitForElement(() => getByTestId('repo-list'))
    expect(repoList).toHaveTextContent(getReposSuccess[0].name)
    expect(repoList).toHaveTextContent(getReposSuccess[0].open_issues_count)
  })

  it('should display a message if the user does not have any repos', async () => {
    // Configure Axios to mock a successful call for retrieving repos
    const url = `/users/${TEST_USER}/repos?type=all&sort=updated`
    const axiosMock = new MockAdapter(axios)
    const mockData = []
    axiosMock.onGet(url).reply(200, mockData)

    // Set the intial state on the store to preload a username
    gitHubStore.currentUser = TEST_USER

    const getReposSpy = jest.spyOn(gitHubStore, 'getRepos')

    const { getByTestId } = render(
      <Provider {...stores}>
        <IntlProvider locale="en">
          <GitHubStoreContext.Provider value={gitHubStore}>
            <HomePage />
          </GitHubStoreContext.Provider>
        </IntlProvider>
      </Provider>,
    )

    const input = document.getElementById('searchTerm')
    expect(input.value).toBe(TEST_USER)
    expect(getReposSpy).toHaveBeenCalled()

    // Ensure that the loading indicator is shown
    const loadingNode = await waitForElement(() => getByTestId('repo-list-loading'))
    expect(loadingNode)

    // Ensure that the repo list error is shown
    const repoList = await waitForElement(() => getByTestId('repo-list-not-found'))
    expect(repoList).toHaveTextContent(
      'The user could not be found or does not have any repositories.',
    )
  })

  it('should display an error message if repos cannot be loaded', async () => {
    // Configure Axios to mock a successful call for retrieving repos
    const url = `/users/${TEST_USER}/repos?type=all&sort=updated`
    const axiosMock = new MockAdapter(axios)
    const mockData = getReposSuccess
    axiosMock.onGet(url).reply(500, mockData)

    // Set the intial state on the store to preload a username
    gitHubStore.currentUser = TEST_USER

    const getReposSpy = jest.spyOn(gitHubStore, 'getRepos')

    const { getByTestId } = render(
      <Provider {...stores}>
        <IntlProvider locale="en">
          <GitHubStoreContext.Provider value={gitHubStore}>
            <HomePage />
          </GitHubStoreContext.Provider>
        </IntlProvider>
      </Provider>,
    )

    const input = document.getElementById('searchTerm')
    expect(input.value).toBe(TEST_USER)
    expect(getReposSpy).toHaveBeenCalled()

    // Ensure that the loading indicator is shown
    const loadingNode = await waitForElement(() => getByTestId('repo-list-loading'))
    expect(loadingNode)

    // Ensure that the repo list error is shown
    const repoList = await waitForElement(() => getByTestId('repo-list-error'))
    expect(repoList)
  })
})
