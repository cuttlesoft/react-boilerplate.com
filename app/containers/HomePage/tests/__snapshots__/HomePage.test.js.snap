// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should render and match the snapshot 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 eTWpSM"
>
  <div
    class="StyledBox-sc-13pk1d4-0 gVSbfC"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 QJUym"
    >
      <img
        class="StyledImage-ey4zx9-0 cHpGKY"
        src="IMAGE_MOCK"
        style="width: 300px;"
      />
    </div>
    <h2
      class="StyledHeading-sc-1rdh4aw-0 jNKvMz"
      style="font-family: soleil, 'Helvetica Neue', Helvetica, Arial, sans-serif; max-width: 100vw; text-transform: false;"
    >
      <span>
        LiveRider
      </span>
    </h2>
    <span
      class="StyledText-sc-1sadyjn-0 hUokoe Text__StyledGrommetText-sc-1qclgcc-0 lkunaY"
    >
      <span>
        The link between music venues, touring artists, and crew.
      </span>
    </span>
    <div
      class="StyledBox-sc-13pk1d4-0 hBDGhc"
    >
      <button
        class="StyledButton-sc-323bzc-0 dzMTYh Button__StyledGrommetButton-sc-5cxguj-0 liIZVq"
        type="button"
      >
        Login
      </button>
      <div
        class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ffjnxm"
      />
      <button
        class="StyledButton-sc-323bzc-0 dzMTYh Button__StyledGrommetButton-sc-5cxguj-0 liIZVq"
        type="button"
      >
        Register
      </button>
    </div>
  </div>
</div>
`;
